adcc/LazyMp.py

```diff
@@ -179,3 +232,105 @@ def mp2_dipole_moment(self):
         mp2corr = -np.array([product_trace(comp, self.mp2_diffdm)
                              for comp in dipole_integrals])
         return refstate.dipole_moment + mp2corr
+
+
+def compute_t2(eri, df1, df2):
+    """Computes T2 amplitudes given an integral block
+    and two delta-Fock-like matrices for the denominator
+    """
+    return evaluate(
+        eri / direct_sum("ia+jb->ijab", df1, df2).symmetrise((2, 3))
+    )
+
+
+def compute_energy_correction(eri, t2):
+    """Computes MP energy corrections
+    by contracting an integral block with a T2 amplitude
+    """
+    return -0.25 * einsum('ijab,ijab->', eri, t2)
+
+
+def compute_mp2_diffdm(hf, mp, space, apply_cvs=None):
+    """Expressions to compute the MP2 difference density matrix
+
+    If `apply_cvs` is `False` then the CVS approximation will not be
+    applied during the density matrix computation even if it is
+    applied in the hf reference. This is done in order to be
+    able to obtain the *full* MP2 difference density (needed for
+    MP properties and MP energies), while keeping the CVS is needed
+    for the CVS-ADC(3) matrix for example.
+    """
+    if apply_cvs is None:
+        apply_cvs = hf.has_core_occupied_space
+
+    # TODO: too much if/else?
```

Comment:
Absolutely.

I think there are two options and you have to see which one is better: Either you write the expressions to a dict (i.e.
```
expressions[b.oo] = ...
if cvs:
    expressions[b.oo] = ...
```
) or you have separate small functions for them and then dispatch. Perhaps also a mixture could be reasonable (one function for keeping cvs one for dropping it and inside you have the expression dicts.

Reply:
Resolved.